helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -627,26 +627,30 @@ private RollingUpdateTaskResult rollingUpdateAwaitRunning(final DeploymentGroup
     final Map<JobId, TaskStatus> taskStatuses = getTaskStatuses(client, host);
 
     if (!taskStatuses.containsKey(deploymentGroup.getJobId())) {
-      // job hasn't shown up yet, probably still being written
+      // Handle cases where agent has not written job status to zookeeper.
+
+      // If job is not listed under /config/hosts node, it was probably deployed successfully and
+      // then manually undeployed. The job will not get redeployed, so treat this as a failure.
+      final Deployment deployment = getDeployment(host, deploymentGroup.getJobId());
+      if (deployment == null) {
+        return RollingUpdateTaskResult.error("job was manually undeployed");
```

Comment:
There are fringe cases where the "manual" part of this statement is not quite true. E.g. if someone created an overlapping deployment group and used rolling-update with the `--migrate` flag.

Maybe something along the lines of:

"Job unexpectedly undeployed on host %s. Perhaps it was undeployed manually?"